[CIR] Upstream overflow builtins

adams381 opened the pull request:
This implements the builtins that handle overflow.

This fixes issue https://github.com/llvm/llvm-project/issues/163888

adams381 committed:
[CIR] Upstream overflow builtins

This implements the builtins that handle overflow.

xlauko reviewed (commented):
Hi @adams381 thank you for your first contribution. I will have few general notes on your PR. As we are upstreaming incubator code we are alining it more with upstream clang code standards. That is for instance: 
- use of [auto](https://llvm.org/docs/CodingStandards.html#use-auto-type-deduction-to-make-code-more-readable). General rule of thumb is to use it only in case the type is explicitly alredy used on the line, e.g. `auto x = cast<SomeType>(y)`
- lower camelCase for variable names

Can you please update the PR before moving forward.

xlauko commented on clang/lib/CIR/CodeGen/CIRGenBuiltin.cpp:
```diff
@@ -58,6 +58,52 @@ static RValue emitBuiltinBitOp(CIRGenFunction &cgf, const CallExpr *e,
+    if (Width < MinWidth) {
+      Width = MinWidth;
+    }
+  }
```
```suggestion
unsigned Width = 0;
for (const auto &Type : Types)
  Width = std::max(Width, Type.Width + (Signed && !Type.Signed));
```

xlauko commented on clang/lib/CIR/CodeGen/CIRGenBuiltin.cpp:
```diff
@@ -58,6 +58,52 @@ static RValue emitBuiltinBitOp(CIRGenFunction &cgf, const CallExpr *e,
+  bool Signed = false;
+  for (const auto &Type : Types) {
+    Signed |= Type.Signed;
+  }
```
```suggestion
bool Signed = llvm::any_of(Types, [](const auto &T) { return T.Signed; });
```

xlauko commented on clang/lib/CIR/Lowering/DirectToLLVM/LowerToLLVM.cpp:
```diff
@@ -2503,6 +2503,118 @@ mlir::LogicalResult CIRToLLVMCmpOpLowering::matchAndRewrite(
+      rhs = rewriter.create<mlir::LLVM::SExtOp>(loc, encompassedLLVMTy, rhs);
+    } else {
+      lhs = rewriter.create<mlir::LLVM::ZExtOp>(loc, encompassedLLVMTy, lhs);
+      rhs = rewriter.create<mlir::LLVM::ZExtOp>(loc, encompassedLLVMTy, rhs);
```
rewriter.create is deprecated construct, use `Op::create(rewriter,` instead

xlauko commented on clang/test/CIR/CodeGen/builtins-overflow.cpp:
Update to follow test structure of other CIR tests:
- add LLVM check of expected LLVM
- add OGCG checks of expected llvm output of original codegen that should match through CIR produced LLVM

andykaylor commented on clang/include/clang/CIR/Dialect/Builder/CIRBaseBuilder.h:
```diff
@@ -408,6 +408,20 @@ class CIRBaseBuilderTy : public mlir::OpBuilder {
+    mlir::Value overflow;
+  };
+
+  BinOpOverflowResults createBinOpOverflowOp(mlir::Location loc,
```
I think it would be better to eliminate this function and just call `cir::BinOpOverflowOp::create` directly where it is currently used. The added `BinOpOverflowResults` structure feels less natural than just calling `getResult` and `getOverflow` on the returned operation.

andykaylor commented on clang/lib/CIR/CodeGen/CIRGenBuiltin.cpp:
```diff
@@ -58,6 +58,52 @@ static RValue emitBuiltinBitOp(CIRGenFunction &cgf, const CallExpr *e,
+namespace {
+struct WidthAndSignedness {
+  unsigned Width;
+  bool Signed;
```
```suggestion
  unsigned width;
  bool signed;
```

andykaylor commented on clang/lib/CIR/CodeGen/CIRGenBuiltin.cpp:
```diff
@@ -58,6 +58,52 @@ static RValue emitBuiltinBitOp(CIRGenFunction &cgf, const CallExpr *e,
+getIntegerWidthAndSignedness(const clang::ASTContext &astContext,
+                             const clang::QualType Type) {
+  assert(Type->isIntegerType() && "Given type is not an integer.");
+  unsigned Width = Type->isBooleanType()  ? 1
```
```suggestion
  unsigned width = type->isBooleanType()  ? 1
```

andykaylor commented on clang/lib/CIR/CodeGen/CIRGenBuiltin.cpp:
```diff
@@ -58,6 +58,52 @@ static RValue emitBuiltinBitOp(CIRGenFunction &cgf, const CallExpr *e,
+  unsigned Width = Type->isBooleanType()  ? 1
+                   : Type->isBitIntType() ? astContext.getIntWidth(Type)
+                                          : astContext.getTypeInfo(Type).Width;
+  bool Signed = Type->isSignedIntegerType();
```
```suggestion
  bool signed = Type->isSignedIntegerType();
```

andykaylor commented on clang/lib/CIR/CodeGen/CIRGenBuiltin.cpp:
```diff
@@ -491,6 +537,154 @@ RValue CIRGenFunction::emitBuiltinExpr(const GlobalDecl &gd, unsigned builtinID,
+      break;
+    }
+
+    auto Loc = getLoc(e->getSourceRange());
```
```suggestion
    mlir::Location loc = getLoc(e->getSourceRange());
```
We don't use auto unless the type is obvious from the code (such as the result of `mlir::cast` or `T::get`) or unhelpfully verbose (such as iterators).

https://llvm.org/docs/CodingStandards.html#use-auto-type-deduction-to-make-code-more-readable

andykaylor commented on clang/lib/CIR/CodeGen/CIRGenBuiltin.cpp:
```diff
@@ -491,6 +537,154 @@ RValue CIRGenFunction::emitBuiltinExpr(const GlobalDecl &gd, unsigned builtinID,
+        ResultArg->getType()->castAs<clang::PointerType>()->getPointeeType();
+    auto ResultCIRTy = mlir::cast<cir::IntType>(cgm.convertType(ResultQTy));
+
+    auto Loc = getLoc(e->getSourceRange());
```
No auto here

andykaylor commented on clang/lib/CIR/Lowering/DirectToLLVM/LowerToLLVM.cpp:
```diff
@@ -2503,6 +2503,118 @@ mlir::LogicalResult CIRToLLVMCmpOpLowering::matchAndRewrite(
+  auto intrinRetTy = mlir::LLVM::LLVMStructType::getLiteral(
+      rewriter.getContext(), {encompassedLLVMTy, overflowLLVMTy});
+
+  auto callLLVMIntrinOp = rewriter.create<mlir::LLVM::CallIntrinsicOp>(
```
```suggestion
  auto callLLVMIntrinOp = mlir::LLVM::CallIntrinsicOp::create(rewriter,
```
This code was using an old idiom that is now deprecated. Similarly for the `create` calls below.

andykaylor commented on clang/lib/CIR/Lowering/DirectToLLVM/LowerToLLVM.cpp:
```diff
@@ -2503,6 +2503,118 @@ mlir::LogicalResult CIRToLLVMCmpOpLowering::matchAndRewrite(
+    }
+  }
+
+  auto intrinName = getLLVMIntrinName(arithKind, encompassedTyInfo.sign,
```
Wow! We really used `auto` a lot in this function. I'll just leave it to the rule I cited above for the rest of them. Some will be ok, others not.

adams381 committed:
Update clang/lib/CIR/CodeGen/CIRGenBuiltin.cpp

Co-authored-by: Henrich Lauko <henrich.lau@gmail.com>

adams381 committed:
Update clang/lib/CIR/CodeGen/CIRGenBuiltin.cpp

Co-authored-by: Henrich Lauko <henrich.lau@gmail.com>

adams381 committed:
Update clang/lib/CIR/CodeGen/CIRGenBuiltin.cpp

Co-authored-by: Andy Kaylor <akaylor@nvidia.com>

adams381 committed:
Update clang/lib/CIR/CodeGen/CIRGenBuiltin.cpp

Co-authored-by: Andy Kaylor <akaylor@nvidia.com>

adams381 committed:
[CIR] Remove createBinOpOverflowOp helper function

Remove the BinOpOverflowResults struct and createBinOpOverflowOp helper
function from CIRBaseBuilder. Instead, call cir::BinOpOverflowOp::create
directly and use getResult() and getOverflow() on the returned operation.

This simplifies the API and makes it more natural to use, as suggested
by reviewer feedback.

adams381 committed:
Update clang/lib/CIR/CodeGen/CIRGenBuiltin.cpp

Co-authored-by: Andy Kaylor <akaylor@nvidia.com>

adams381 committed:
Update clang/lib/CIR/CodeGen/CIRGenBuiltin.cpp

Co-authored-by: Andy Kaylor <akaylor@nvidia.com>

adams381 committed:
Update clang/lib/CIR/Lowering/DirectToLLVM/LowerToLLVM.cpp

Co-authored-by: Andy Kaylor <akaylor@nvidia.com>

adams381 committed:
[CIR] Fix style and deprecated API in overflow builtins lowering

Replace auto with explicit types and use lowerCamelCase.
Replace deprecated rewriter.create with Op::create.

adams381 committed:
[CIR] Fix style and remove helper function in overflow builtins

- Rename 'signed' field to 'isSigned' (signed is a keyword)
- Use lowerCamelCase for all variables and parameters
- Replace createBinOpOverflowOp helper with direct BinOpOverflowOp::create calls

andykaylor commented on clang/test/CIR/CodeGen/builtins-overflow.cpp:
```diff
@@ -0,0 +1,364 @@
+// RUN: %clang_cc1 -triple x86_64-unknown-linux-gnu -fclangir -emit-cir %s -o %t.cir
+// RUN: FileCheck %s --check-prefix=CIR --input-file=%t.cir
```
```suggestion
// RUN: %clang_cc1 -triple x86_64-unknown-linux-gnu -fclangir -emit-cir %s -o %t.cir
// RUN: FileCheck %s --check-prefix=CIR --input-file=%t.cir
// RUN: %clang_cc1 -triple x86_64-unknown-linux-gnu -fclangir -emit-llvm %s -o %t-cir.ll
// RUN: FileCheck %s --check-prefix=LLVM --input-file=%t-cir.ll
// RUN: %clang_cc1 -triple x86_64-unknown-linux-gnu -emit-llvm %s -o %t.ll
// RUN: FileCheck %s --check-prefix=OGCG --input-file=%t.ll
```
Can you add LLVM and OGCG checks based on the above run lines?

andykaylor reviewed (commented):
It looks like some of your changes may not have made it into the update.

xlauko commented on clang/lib/CIR/Lowering/DirectToLLVM/LowerToLLVM.cpp:
```diff
@@ -2503,6 +2503,125 @@ mlir::LogicalResult CIRToLLVMCmpOpLowering::matchAndRewrite(
+CIRToLLVMBinOpOverflowOpLowering::EncompassedTypeInfo
+CIRToLLVMBinOpOverflowOpLowering::computeEncompassedTypeWidth(
+    cir::IntType operandTy, cir::IntType resultTy) {
+  auto sign = operandTy.getIsSigned() || resultTy.getIsSigned();
```
```suggestion
  bool sign = operandTy.getIsSigned() || resultTy.getIsSigned();
```

xlauko commented on clang/lib/CIR/Lowering/DirectToLLVM/LowerToLLVM.cpp:
```diff
@@ -2503,6 +2503,125 @@ mlir::LogicalResult CIRToLLVMCmpOpLowering::matchAndRewrite(
+                                                  encompassedLLVMTy, truncResult);
+    mlir::Value truncOverflow =
+        mlir::LLVM::ICmpOp::create(rewriter, loc, mlir::LLVM::ICmpPredicate::ne,
+                                   truncResultExt, result);
```
```suggestion
    auto truncOverflow =
        mlir::LLVM::ICmpOp::create(rewriter, loc, mlir::LLVM::ICmpPredicate::ne,
                                   truncResultExt, result);
```

adams381 committed:
Update clang/lib/CIR/CodeGen/CIRGenBuiltin.cpp

Co-authored-by: Andy Kaylor <akaylor@nvidia.com>

adams381 committed:
Update clang/test/CIR/CodeGen/builtins-overflow.cpp

Co-authored-by: Andy Kaylor <akaylor@nvidia.com>

adams381 committed:
Update clang/include/clang/CIR/Dialect/IR/CIROps.td

Co-authored-by: Henrich Lauko <henrich.lau@gmail.com>

adams381 committed:
Update clang/lib/CIR/Lowering/DirectToLLVM/LowerToLLVM.cpp

Co-authored-by: Henrich Lauko <henrich.lau@gmail.com>

adams381 committed:
Update clang/lib/CIR/Lowering/DirectToLLVM/LowerToLLVM.cpp

Co-authored-by: Henrich Lauko <henrich.lau@gmail.com>

adams381 committed:
Update clang/lib/CIR/Lowering/DirectToLLVM/LowerToLLVM.cpp

Co-authored-by: Henrich Lauko <henrich.lau@gmail.com>

adams381 committed:
Update clang/lib/CIR/Lowering/DirectToLLVM/LowerToLLVM.cpp

Co-authored-by: Henrich Lauko <henrich.lau@gmail.com>

adams381 committed:
Update clang/lib/CIR/Lowering/DirectToLLVM/LowerToLLVM.cpp

Co-authored-by: Henrich Lauko <henrich.lau@gmail.com>

adams381 committed:
Update clang/lib/CIR/Lowering/DirectToLLVM/LowerToLLVM.cpp

Co-authored-by: Henrich Lauko <henrich.lau@gmail.com>

adams381 committed:
Update clang/lib/CIR/Lowering/DirectToLLVM/LowerToLLVM.cpp

Co-authored-by: Henrich Lauko <henrich.lau@gmail.com>

adams381 committed:
[CIR] Fix style: lowercase variables and clang-format

- Convert all variable names to lowerCamelCase per LLVM coding standards
- Fix syntax error: auto mlir::Value -> auto
- Apply clang-format for consistent 80-column line wrapping
- Keep auto usage only where type is obvious from same line

adams381 committed:
[CIR] Fix BinOpOverflowOp assembly format for multiple results

The assembly format for operations with multiple results requires a comma
separator, not an arrow. Changed from:
  : qualified(type($lhs)) ->
to:
  : qualified(type($lhs)) ,

This produces the correct output format:
  : !u32i, (!u32i, !cir.bool)
instead of the incorrect:
  : !u32i -> (!u32i, !cir.bool)

Also added qualified() to properly format bit-precise integer types as
!cir.int<s, 31> instead of <s, 31>.

Added LLVM and OGCG checks to the test to verify CIR lowering produces
correct LLVM intrinsics matching the original Clang codegen.

andykaylor committed:
Merge branch 'main' into overflow_builtins

andykaylor committed:
Fix formatting
